typespec-ccf/proposals.tsp

```diff
@@ -0,0 +1,203 @@
+  @visibility("read")
+  @maxLength(64)
+  @format("hex")
+  proposerId: string;
+
+  @doc("Current state of this proposal.")
+  proposalState: ProposalState;
+
+  @doc("Count of how many ballots have been submitted for this proposal.")
+  ballotCount: uint64;
+
+  @doc("If a proposal is not open, then this contains the result of each ballot when the proposal transitioned from open.")
+  finalVotes?: MemberVotes;
+
+  @doc("This will be populated with a description of any ballots which failed to execute.")
+  voteFailures?: VoteFailures;
+
+  @doc("If proposalState is Failure, then this will describe the proposal's failure.")
+  failure?: FailureInfo;
+}
+
+@doc("The actions contained in a proposal, describing the proposal's changes.")
+@TypeSpec.Rest.parentResource(Proposal)
+model ProposalActions {
+  @doc("A list of actions to apply. Each action is considered, in order, for both proposal and execution of the proposal. All actions are validated and, if the proposal is accepted, applied atomically.")
+  actions: Action[];
+}
+
+@doc("The source script of an executable vote from a member, regarding a proposed change to the service.")
+@TypeSpec.Rest.resource("ballots")
+@TypeSpec.Rest.parentResource(Proposal)
+model Ballot {
+  @doc("ID of CCF member who signed and submitted this ballot.")
+  @visibility("read")
+  @maxLength(64)
+  @format("hex")
+  @key
+  memberId: string;
+
+  @doc("The script to execute, returning the voter's support of this proposal.")
+  @visibility("create")
+  ballot: string;
+}
+
+@doc("A compact summary of the service's state up to a certain point in time, updated and signed by members to indicate their participation in and approval of the service.")
+@TypeSpec.Rest.resource("stateDigests")
+model StateDigest {
+  @doc("Identifier for member this stateDigest applies to.")
+  @visibility("read")
+  @maxLength(64)
+  @format("hex")
+  @key
+  memberId: string;
+
+  @doc("Hex-encoding of SHA-256 hash of the root of the service's merkle tree. This should be signed by a new member and submitted as an ACK to mark that member as Active.")
+  @maxLength(64)
+  @format("hex")
+  stateDigest: string;
+}
+
+model CoseSigned<T = {}> {
+  @doc("Content-type for all auditable member actions is application/cose, containing a CoseSign1 signature.")
+  @TypeSpec.Http.header("Content-type")
+  contentType: "application/cose";
+
+  @doc("A raw CoseSign1 signature. The required protected headers and content schema will vary by action.")
+  @TypeSpec.Http.body
+  body: bytes;
+}
+
+interface Proposals {
+  @doc("Submit a proposed change to the service. This will be assigned an ID by the service. Idempotence is based on the signed proposal body - submitting identical signed proposals multiple times will only create a single proposal on the server, unless sufficient time has passed (since the `created_at` signed COSE header), in which case the submission is stale and will return an error.")
```

Comment:
Wonder if we can clarify this somewhat. It is not obviously apparent that the exact same _proposal_ is not actually identical when the signing timestamps are different.

Reply:
e.g. `submitting identical signed proposals (including timestamp headers)`

Reply:
Good clarification. By "signed proposal body", I meant the request body, but realise now that's ambiguous (could be the Cose payload, not the entire CoseSign1). I think the right thing to do here is to link to the CCF docs, where we can write a more thorough (and updatable) explanation of proposal identity + idempotence.

Reply:
I've tweaked the string slightly, and created a TODO to get a clearer docs page to link to:

https://github.com/microsoft/CCF/issues/5230